src/jit/morph.cpp

```diff
@@ -12501,14 +12502,16 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                 lclVarTree->gtFlags |= GTF_VAR_DEF;
             }
 
-            if (op1->gtEffectiveVal()->OperIsConst())
+            effectiveOp1 = op1->gtEffectiveVal();
+
+            if (effectiveOp1->OperIsConst())
```

Comment:
In theory `effectiveOp1` should be updated inside this `if`. In practice what this `if` does is beyond weird. It's either some legacy thing or it's completely bogus. Either way, I've chosen not to update `effectiveOp1` so that the cast doesn't get removed in some weird scenario that can't be reproduced and tested.